feat: add config option for mount method hostpath or virtual device (#2459)
odigos-io/odigos: docs/instrumentations/configuration/mount-method.mdx

---

> title: "Agent Mount Method"

Reviewer: What do you think about:
1. `Agent Injection Mode` 
2. `Instrumentation Mode`

Author: I think injection and instrumentaiton can mean many things. mount makes what this option is for explicit.

I am open to hearing alternatives

Reviewer: I thought to abstract it for the user, basically the mount used to inject the agent for languages that needed.
I feel like Agent mount is not clear, but maybe lets hear another opinion

sidebarTitle: "Mount Method"
---

## Mount Method

> For odigos agents to run in the instrumented pod containers, some files need to be mounted to the container.

Reviewer: ```suggestion
For Odigos agents to run inside instrumented pod containers, certain files must be mounted into the container.
```

In Kubernetes, these files are mounted under `/var/odigos` directory, with subdirectories for each agent.

<Warning>
This section is for advanced users and odigos administrators.

It is recommended to use the default settings, unless you have specific requirements or run into an issue.
</Warning>

## Supported Mount Methods

Odigos supports 2 mount methods, which can be used depending on the user preference, cluster policies, integration with existing tools, etc.

> It is recommended to use `HostPath` unless it does not work, in this case fallback to `VirtualDevice`.

Reviewer: I think redundant, our recommendation written in the HostPath maybe bold it ? :)

Author: agree. removed

### HostPath

This is the default and the recommended method.

#### Pod Manifest Additions

For agents and languages that requires fs mounts, odigos webhook will add the following:

- Volume to the pod spec, under the `spec.volumes` field:

```yaml
  - hostPath:
      path: /var/odigos
      type: ""
    name: odigos-agent
```

- VolumeMount to the instrumented container specs, under the `spec.containers[].volumeMounts` field:

```yaml
    - mountPath: /var/odigos/{agent_sub_dir}
      name: odigos-agent
      readOnly: true
      subPath: {agent_sub_dir}
```

#### Caveats

> The "host path" option should be enabled in the cluster. Some policy tools may block this option, like: 

Reviewer: ```suggestion
The "HostPath" option should be enabled in the cluster. Some policy tools may block this option, like: 
```

- [Open Policy Agent](https://www.openpolicyagent.org/)
- [Kyverno](https://kyverno.io/)

If you cluster has such policies, you can:

- use the `VirtualDevice` method, or

> - ask the cluster admin to whitelist the odigos agent host path in the policy tool.

Reviewer: If your cluster enforces such policies, you have the following options:
- Use the VirtualDevice method.
- Request the cluster administrator to whitelist the Odigos agent host path in the policy tool.

Author: like 👍  updated

---
title: "Agent Mount Method"
sidebarTitle: "Mount Method"

> ---

Reviewer 2: Please add an icon, use the icon names from: https://fontawesome.com/icons

## Mount Method

> For Odigos agents to run inside instrumented pod containers, certain files must be mounted into the container.

Reviewer 2: What do you think about adding this to the "description" on the page metadata above?

Author: good idea, moved this text there

In Kubernetes, these files are mounted under `/var/odigos` directory, with subdirectories for each agent.

> <Warning>

Reviewer 2: I think this needs to be "Info", like in [introduction](https://docs.odigos.io/quickstart/introduction)

Author: I intentionally wanted it to be warning so it's clear normal users should not care about these settings

This section is for advanced users and odigos administrators.

It is recommended to use the default settings, unless you have specific requirements or run into an issue.
</Warning>

## Supported Mount Methods

Odigos supports 2 mount methods, which can be used depending on the user preference, cluster policies, integration with existing tools, etc.

> ### VirtualDevice

Reviewer 2: Make this a list please.
The title "Supported Mount Methods" and it's description suggests this should be numbered.

Author: good idea, added numbers

<Info>
This is the default mount method
</Info>

#### Pod Manifest Additions

> For agents and languages that requires fs mounts, odigos webhook will add the following:

Reviewer 2: "fs mounts" has not been defined before this line, one can assume it's "file system", but this should be clear without assumptions.

Author: changed to filesystem 👍

- For each instrumented container in the pod spec, odigos will add a resource requirement under the `spec.containers[].resources` field:

```yaml
    resources:
      limits:
        instrumentation.odigos.io/generic: "1"
      requests:
        instrumentation.odigos.io/generic: "1"
```

### Caveats

- May sometimes not integrate well with node auto-scaling tools like [Karpenter](https://karpenter.sh/).
- Odiglet daemonset needs to be running on a node for instrumented pods to be scheduled on that node.

### HostPath

This method is an opt-in configuration option which can be used if the virtual device method is not suitable for your cluster.

If it is supported in the cluster, it is preferred to use over the VirtualDevice method which requires odiglet component to run.

> #### Enabling HostPath

Reviewer 2: nit, but I would prefer this syntax:
```
**Enabling HostPath**
```

It gives the boldness, without changing font size, perfect for subtitles under a 3x hashtag (### HostPath)

Author: replaced with this bold text

> #### Enabling VirtualDevice

Reviewer 3: ```suggestion
#### Enabling HostPath
```

Author: thanks, fixed

- Profile: `odigos profile add mount-method-k8s-host-path`

- Odigos CLI: `odigos config set mount-method k8s-host-path`

- Helm Chart: in your values file, set `instrumentor.mountMethod` to `k8s-host-path`, or use helm cli `--set instrumentor.mountMethod=k8s-host-path` flag with helm upgrade.

- Kubernetes Manifest: under the `odigos-config` ConfigMap in odigos namespace, set the value in the `mountMethod` field of `config.yaml` to `k8s-host-path`.

> #### Pod Manifest Additions

Reviewer 3: Should this be un-commented?

Author: It's not commented, it's the markdown header format

For agents and languages that requires filesystem mounts, odigos webhook will add the following:

- Volume to the pod spec, under the `spec.volumes` field:

```yaml
  - hostPath:
      path: /var/odigos
      type: ""
    name: odigos-agent
```

- VolumeMount to the instrumented container specs, under the `spec.containers[].volumeMounts` field:

```yaml
    - mountPath: /var/odigos/{agent_sub_dir}
      name: odigos-agent
      readOnly: true
      subPath: {agent_sub_dir}
```

> #### Caveats

Reviewer 2: Same here, is this a child of "HostPath"?
The please use stars:
```
**Caveats**
```
